azure/iac/Redis.rego

```diff
@@ -142,4 +142,208 @@ serverRole_metadata := {
+}
+
+
+# PR-AZR-ARM-ARC-005
+#
+
+default arc_subnet_id = null
+
+azure_attribute_absence["arc_subnet_id"] {
+    resource := input.resources[_]
+    lower(resource.type) == "microsoft.cache/redis"
+    not resource.properties.subnetId
+}
+
+source_path[{"arc_subnet_id":metadata}] {
+    resource := input.resources[i]
+    lower(resource.type) == "microsoft.cache/redis"
+    not resource.properties.subnetId
+    metadata:= {
+        "resource_path": [["resources",i,"properties","subnetId"]]
+    }
+}
+
+
+azure_issue["arc_subnet_id"] {
+    resource := input.resources[_]
+    lower(resource.type) == "microsoft.cache/redis"
+    count(resource.properties.subnetId) == 0
+}
+
+
+source_path[{"arc_subnet_id":metadata}] {
+    resource := input.resources[i]
+    lower(resource.type) == "microsoft.cache/redis"
+    count(resource.properties.subnetId) == 0
+    metadata:= {
+        "resource_path": [["resources",i,"properties","subnetId"]]
+    }
+}
+
+arc_subnet_id {
+    lower(input.resources[_].type) == "microsoft.cache/redis"
+    not azure_issue["arc_subnet_id"]
+    not azure_attribute_absence["arc_subnet_id"]
+}
+
+arc_subnet_id = false {
+    azure_issue["arc_subnet_id"]
+}
+
+arc_subnet_id = false {
+    azure_attribute_absence["arc_subnet_id"]
+}
+
+arc_subnet_id_err = "Azure Cache for Redis is not reside within a virtual network" {
+    azure_issue["arc_subnet_id"]
+} else = "Azure Cache for Redis attribute subnetId missing in the resource" {
+    azure_attribute_absence["arc_subnet_id"]
+}
+
+
+arc_subnet_id_metadata := {
+    "Policy Code": "PR-AZR-ARM-ARC-005",
+    "Type": "IaC",
+    "Product": "AZR",
+    "Language": "ARM template",
+    "Policy Title": "Azure Cache for Redis should use private link",
+    "Policy Description": "Azure Virtual Network deployment provides enhanced security and isolation for your Azure Cache for Redis, as well as subnets, access control policies, and other features to further restrict access.When an Azure Cache for Redis instance is configured with a virtual network, it is not publicly addressable and can only be accessed from virtual machines and applications within the virtual network.",
+    "Resource Type": "microsoft.cache/redis",
+    "Policy Help URL": "",
+    "Resource Help URL": "https://docs.microsoft.com/en-us/azure/templates/microsoft.cache/redis"
+}
+
+#
+# PR-AZR-ARM-ARC-006
+
+default arc_private_endpoint = null
+
+azure_attribute_absence["arc_private_endpoint"] {
+    count([c | lower(input.resources[_].type) == "microsoft.cache/redis"; c := 1]) != count([c | lower(input.resources[_].type) == "microsoft.network/privateendpoints"; c := 1])
+}
+
+azure_issue["arc_private_endpoint"] {
+    resource := input.resources[_]
+    lower(resource.type) == "microsoft.network/privateendpoints"
+    privateLinkServiceConnection := resource.properties.privateLinkServiceConnections[_]
+    contains(lower(privateLinkServiceConnection.properties.privateLinkServiceId), "microsoft.cache/redis")
```

Comment:
**there should be a not before contains.**
We need to check combination of **resource type and name** contains in the `privateLinkServiceId `instead of checking only the type. otherwise if there are multiple redis cache we wont be able to identify which redis cache has link with `microsoft.network/privateendpoints`

**note: a valid id always looks like:** `/subscriptions/{subs_id}/resourceGroups/{rg_id}/providers/Microsoft.Cache/Redis/dev-redis-cache`.
string after the last forward slash '/' is the name of redis cache.
So if we need to find a link between private endpoint and redis cache, we need to check if `privateLinkServiceId `contains `'Microsoft.Cache/Redis/dev-redis-cache'`

**Also please file an issue if snapshot is not generating the type and name out of variable in privateLinkServiceId** 

**Code sample:**

```
azure_issue ["arc_private_endpoint"] {
    resource := input.resources[_]
    lower(resource.type) == "microsoft.network/privateendpoints"
    privateLinkServiceConnection := resource.properties.privateLinkServiceConnections[_]

    count([c | r := input.resources[_];
              lower(r.type) == "microsoft.cache/redis";
              contains(lower(privateLinkServiceConnection.properties.privateLinkServiceId), lower(concat("/", [r.type, r.name])));
              c := 1]) == 0
}
```

Reply:
ignore previous comment and add a `not `before `contains`

Reply:
@shahinshirvani change the function like this. its safer (we can discuss if you have question):

azure_issue ["arc_private_endpoint"] {
    resource := input.resources[_]
    lower(resource.type) == "microsoft.cache/redis"
    count([c | r := input.resources[_];
              lower(r.type) == "microsoft.network/privateendpoints";
              contains(lower(r.properties.privateLinkServiceConnections[_].properties.privateLinkServiceId), lower(resource.type));
              c := 1]) == 0
}

Reply:
Both are correct existing one and the suggested one. can use anyone.